## Ejercicio - Bases de Datos No Relacionales con Mongo

#### ○ Conexión a la Plataforma seleccionada de bases de datos no relacionales
#### ○ Generar una nueva base de datos 

In [2]:
from pymongo import MongoClient

# Requires the PyMongo package.
# https://api.mongodb.com/python/current

client = MongoClient('mongodb+srv://hzoscarist:irxgsjdfOoSHmygn@cluster1.xaxbyn6.mongodb.net/')

In [21]:
print(client.list_database_names())

['sample_airbnb', 'sample_analytics', 'sample_geospatial', 'sample_guides', 'sample_mflix', 'sample_restaurants', 'sample_supplies', 'sample_training', 'sample_weatherdata', 'tweets', 'admin', 'local']


In [23]:
db = client.tweets
col = db.tweets

In [24]:
## col.count_documents({'airline_sentiment':'negative'})

9178

#### ○ Explorar el archivo de tweets con palabras clave definidos para el caso (incluir 5)

##### 1. Se agrupa por la percepción del sentimiento de la aerolínea

In [30]:
agg_result=col.aggregate([
    { "$group": {
    "_id": "$airline_sentiment",
    "count": { "$sum": 1 }
    }}
    ]); 

In [31]:
for i in agg_result:
    print(i)

{'_id': 'neutral', 'count': 3099}
{'_id': 'positive', 'count': 2363}
{'_id': 'negative', 'count': 9178}


##### 2. Se agrupa por el motivo por el cual se tuvo una percepción  negativa de la aerolínea

In [32]:
agg_result=col.aggregate([
    { "$group": {
    "_id": "$negativereason",
    "count": { "$sum": 1 }
    }}
    ]); 

In [33]:
for i in agg_result:
    print(i)

{'_id': None, 'count': 5462}
{'_id': 'Cancelled Flight', 'count': 847}
{'_id': 'Bad Flight', 'count': 580}
{'_id': 'Damaged Luggage', 'count': 74}
{'_id': 'longlines', 'count': 178}
{'_id': 'Flight Booking Problems', 'count': 529}
{'_id': 'Lost Luggage', 'count': 724}
{'_id': 'Late Flight', 'count': 1665}
{'_id': 'Customer Service Issue', 'count': 2910}
{'_id': "Can't Tell", 'count': 1190}
{'_id': 'Flight Attendant Complaints', 'count': 481}


##### 3. Se agrupa por las diferentes aerolíneas

In [34]:
agg_result=col.aggregate([
    { "$group": {
    "_id": "$airline",
    "count": { "$sum": 1 }
    }}
    ]); 

In [35]:
for i in agg_result:
    print(i)

{'_id': 'Southwest', 'count': 2420}
{'_id': 'United', 'count': 3822}
{'_id': 'Delta', 'count': 2222}
{'_id': 'Virgin America', 'count': 504}
{'_id': 'US Airways', 'count': 2913}
{'_id': 'American', 'count': 2759}


##### 4. Se agrupa por las diferentes timezone de los usuarios

In [36]:
agg_result=col.aggregate([
    { "$group": {
    "_id": "$user_timezone",
    "count": { "$sum": 1 }
    }}
    ]); 

In [37]:
for i in agg_result:
    print(i)

{'_id': 'Kyiv', 'count': 2}
{'_id': 'Bucharest', 'count': 1}
{'_id': 'Mazatlan', 'count': 3}
{'_id': 'Seoul', 'count': 5}
{'_id': 'Central America', 'count': 13}
{'_id': 'Bogota', 'count': 5}
{'_id': 'Quito', 'count': 738}
{'_id': 'Central Time (US & Canada)', 'count': 1931}
{'_id': 'America/New_York', 'count': 26}
{'_id': 'Tijuana', 'count': 9}
{'_id': 'Mexico City', 'count': 3}
{'_id': 'Alaska', 'count': 108}
{'_id': 'Wellington', 'count': 1}
{'_id': 'Paris', 'count': 25}
{'_id': 'Brasilia', 'count': 23}
{'_id': 'Stockholm', 'count': 4}
{'_id': 'Kuala Lumpur', 'count': 1}
{'_id': 'Jerusalem', 'count': 3}
{'_id': 'Singapore', 'count': 2}
{'_id': 'America/Los_Angeles', 'count': 15}
{'_id': 'Tokyo', 'count': 1}
{'_id': 'Istanbul', 'count': 1}
{'_id': 'Canberra', 'count': 1}
{'_id': 'America/Atikokan', 'count': 1}
{'_id': 'Brisbane', 'count': 10}
{'_id': 'Perth', 'count': 2}
{'_id': 'Lisbon', 'count': 1}
{'_id': 'Athens', 'count': 16}
{'_id': 'Midway Island', 'count': 1}
{'_id': 'West Ce

##### 5. Se agrupa por la ubicación desde la cual se mando el tweet

In [93]:
agg_result=col.aggregate([
    { "$group": {
    "_id": "$tweet_location",
    "count": { "$sum": 1 }
    }}
    ]); 

In [94]:
## for i in agg_result:
##     print(i)

#### ○ Insights

Despúes de haber hecho todas estas agrupaciones se puede tener una idea del contenido de este dataset:
- Más del 60% de los tweets tuvieron una percepción negativa de la aerolínea en la cual viajaron.

- Aunque hay una gran cantidad de casos en los que no se justifica el motivo, en aquellos casos
en los cuales se menciona se ve que la gran mayoría son por el servicio al cliente. (2910 casos)

- De las 6 aerolíneas consideradas solo hay una que tuvo una cantidad considerablemente inferior
 con respecto a las demás. ("Virgin America" 504) Excluyendo esta última el promedio de los vuelos
 por aerolínea fue de 2827.

 - La gran cantidad de vuelos se encuentran concentrados en América del Norte.

#### ○ Importar por lo menos 1000 tweets a la base de datos a una nueva estructura

- Realicé el filtro por la percepción negativa de la aerolínea a través de los tweets obteniendo más de 9000 pots.

- Exporte dichos pots en un archivo csv que será adjunto en los soportes de esta actividad.

In [95]:
filter={'airline_sentiment':'negative'}

result = client['tweets']['tweets'].find(
  filter=filter
)

In [96]:
## import pprint
## pp = pprint.PrettyPrinter(indent=4)
## for doc in result:
##    pp.pprint(doc)

In [81]:
import csv

csv_columns = ['tweet_id','airline_sentiment','airline_sentiment_confidence', 'negativereason', 'negativereason_confidence', 'airline', 'airline_sentiment_gold', 'name', 'negativereason_gold', 'retweet_count','text','tweet_coord','tweet_created','tweet_location','user_timezone']

with open('negative_sentiment.csv', 'w', encoding='utf-8') as csvfile:
    writer = csv.DictWriter(csvfile, fieldnames=csv_columns,extrasaction='ignore')
    writer.writeheader()
    for post in col.find({'airline_sentiment': 'negative'}):
        writer.writerow(post)

#### ○ Construir 2 filtros de búsqueda que se considere importante dentro de la base de datos de 
#### tweets proporcionada

- El primer filtro corresponde a los tweets que muestran una percepción negativa de una aerolínea
 por el no buen servicio al cliente de una aerolínea en particular.

- Basta cambiar el último filtro para determinar en que difieren, a grandes razgos, los contendios
de los tweets entre una aerolina y otra.

In [97]:
filter={"$and":[
    {"airline_sentiment":"negative"},
    {"negativereason": "Customer Service Issue"},
    {"airline":"United"}]}

result = client['tweets']['tweets'].find(filter=filter)

In [99]:
## import pprint
## pp = pprint.PrettyPrinter(indent=4)
## for doc in result:
##   pp.pprint(doc)

- La segunda query corresponde a los tweets que muestran una percepción negativa de la aerolínea
filtrada por el timezone del usuario que hizo el tweet. Para este caso utilice la ciudad de Bogotá.

- Basta cambiar el último filtro para cambiar el país del cual se quieren obtener los tweets.

In [86]:
filter={"$and":[
    {"airline_sentiment":"negative"},
    {"user_timezone": "Bogota"}]}

result = client['tweets']['tweets'].find(filter=filter)

In [87]:
import pprint
pp = pprint.PrettyPrinter(indent=4)
for doc in result:
    pp.pprint(doc)

{   '_id': ObjectId('64a1d33068a2d70aa5c13ef0'),
    'airline': 'United',
    'airline_sentiment': 'negative',
    'airline_sentiment_confidence': 0.6726,
    'name': 'mariachan90',
    'negativereason': 'Customer Service Issue',
    'negativereason_confidence': 0.3429,
    'retweet_count': 0,
    'text': '@united understanding the situation we waited and it was opened '
            'until 10:30pm',
    'tweet_created': '2015-02-19 20:37:19 -0800',
    'tweet_id': 568630453521022976,
    'user_timezone': 'Bogota'}
{   '_id': ObjectId('64a1d33368a2d70aa5c1513c'),
    'airline': 'Delta',
    'airline_sentiment': 'negative',
    'airline_sentiment_confidence': 1.0,
    'name': 'CamiCorreaBal',
    'negativereason': 'Lost Luggage',
    'negativereason_confidence': 0.6383,
    'retweet_count': 0,
    'text': "@JetBlue yes, We have de baggage claim, I'm so sad for the "
            'baggage and how They treat Us 😞 please we need That baggage',
    'tweet_created': '2015-02-21 10:24:24 -0800'